### PR TITLE
Translatable settings - Rebase of #6873

### PR DESCRIPTION
**Pull request by Tijani-Dia**
This is a rebase of #6873 built on top of #9317.

***Demo***

https://user-images.githubusercontent.com/71412737/196472765-15e16688-afd1-45bd-b5c7-c7eff77d9cf1.mp4

https://user-images.githubusercontent.com/71412737/196474853-a3beea98-bf20-4e08-9cb2-fabb55cf4f0e.mp4

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [ ] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**: Chrome, Firefox
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [x] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

File: wagtail/contrib/settings/tests/translatable_generic/base.py
Comment (zerolab) [inline wagtail/contrib/settings/tests/translatable_generic/base.py:40]: This is great, however it doesn't seem to work with `django.middleware.locale.LocaleMiddleware`. That is, using bakerydemo, and translating a page to say French then visiting that page we get a path like `/fr/the-page/` as the locale is set in `request.LANGUAGE_CODE`

File: wagtail/contrib/settings/views.py
Comment (zerolab) [inline wagtail/contrib/settings/views.py:113]:
```suggestion
    def get_translations(self, locales):
```

File: wagtail/models/i18n.py
Comment (zerolab) [inline wagtail/models/i18n.py:48]: 👌

File: wagtail/contrib/settings/utils.py
Comment (zerolab) [inline wagtail/contrib/settings/utils.py:35]: issue: we cannot rely on locale being passed as a get parameter on the front-end. when using `LocaleMiddleware` and i18n

suggestion: make use of [`Locale.get_active()`](https://github.com/wagtail/wagtail/blob/4c7e1ea1ce4c842ce30905c6539437fdf21be6df/wagtail/models/i18n.py#L69) if a valid locale is not passed as a parameter

File: wagtail/contrib/settings/utils.py
Comment (zerolab) [inline wagtail/contrib/settings/utils.py:45]:
```suggestion
    return Locale.get_default()
```

https://github.com/wagtail/wagtail/blob/4c7e1ea1ce4c842ce30905c6539437fdf21be6df/wagtail/models/i18n.py#L57

File: wagtail/contrib/settings/utils.py
Comment (zerolab) [inline wagtail/contrib/settings/utils.py:38]:
```suggestion
            return locales.get_for_language(selected_locale)
```

https://github.com/wagtail/wagtail/blob/4c7e1ea1ce4c842ce30905c6539437fdf21be6df/wagtail/models/i18n.py#L30

File: wagtail/contrib/settings/models.py
Comment (zerolab) [inline wagtail/contrib/settings/models.py:272]: I feel that this adds too much developer overhead, we are not asking people to use `TranslatablePage` instead of `Page` if they want content to support i18n. And we do offer `TranslatableMixin` for models outside of our control, which is fine.

IMHO, as a developer I should only need to extend `BaseGenericSetting` or `BaseSiteSetting`, and if/when I add i18n, "it should just work" ™️.

**Review by zerolab — CHANGES_REQUESTED** [review]
Thank you for this @Tijani-Dia 

Lots and lots of great stuff, but it needs work.
The main thing here for me is the additional burden on developers with the new settings bases, and the fact that I can't use `{{ settings.my_app.MySettingClass.the_field }}` and get the correct value when using https://docs.wagtail.org/en/stable/advanced_topics/i18n.html#user-language-auto-detection

File: wagtail/contrib/settings/models.py
Comment (Tijani-Dia) [reply wagtail/contrib/settings/models.py:272]: I copied the approach @kaedroho did originally to split the models between (non)/translatable settings. So maybe it'd be nice to know the rationale behind it. Otherwise, I agree with how it'd simplify things for developers.

File: wagtail/contrib/settings/registry.py
Comment (Tijani-Dia) [reply wagtail/contrib/settings/registry.py:54]: Because we don't always get the locale from an instance. We can get it from a request or/and a model: https://github.com/wagtail/wagtail/pull/9390/files#diff-1b8511bbd47182a3cbc4099e5d8426e648c02b2e3b8b3f5d9d1cec6ac61d4a73R55-R58

File: wagtail/contrib/settings/tests/translatable_generic/test_admin.py
Comment (Tijani-Dia) [reply wagtail/contrib/settings/tests/translatable_generic/test_admin.py:34]: To be honest, I copied this from the previous tests. However, it seems that we don't want a superuser, we want a user that has *only* access to the admin.

File: wagtail/contrib/settings/utils.py
Comment (Tijani-Dia) [reply wagtail/contrib/settings/utils.py:56]: Moved `is_default_locale` to the `Locale` model.

Where would you suggest moving `get_locale_for` and `add_locale_query_string`? I thought `wagtail.coreutils`?

File: wagtail/contrib/settings/utils.py
Comment (Tijani-Dia) [reply wagtail/contrib/settings/utils.py]: I agree. I've updated it.

File: wagtail/contrib/settings/utils.py
Comment (Tijani-Dia) [reply wagtail/contrib/settings/utils.py]: Updated in https://github.com/wagtail/wagtail/pull/9390/commits/8f359bc19dd00b6cf6b8ed00cbd0fd2b813dfe85

File: wagtail/contrib/settings/tests/translatable_site_specific/test_admin.py
Comment (zerolab) [reply wagtail/contrib/settings/tests/translatable_site_specific/test_admin.py:27]: resolving on the basis of https://github.com/wagtail/wagtail/pull/9390#discussion_r1002188412

File: wagtail/contrib/settings/tests/generic/test_admin.py
Comment (zerolab) [inline wagtail/contrib/settings/tests/generic/test_admin.py:91]: suggestion: add note as to what those are. I feel that we have a few "magic numbers" like that around.

File: wagtail/contrib/settings/tests/translatable_generic/test_admin.py
Comment (Tijani-Dia) [reply wagtail/contrib/settings/tests/translatable_generic/test_admin.py:261]: That's what I thought too when I first 'copied' it. However, the syntax is valid https://book.pythontips.com/en/latest/for_-_else.html.

File: wagtail/contrib/settings/tests/translatable_generic/test_admin.py
Comment (Tijani-Dia) [reply wagtail/contrib/settings/tests/translatable_generic/test_admin.py]: Addressed in https://github.com/wagtail/wagtail/pull/9390/commits/1c95db83c2ca0dc4440faceb3f63ba5b144698e0

File: docs/reference/contrib/settings.md
Comment (Tijani-Dia) [reply docs/reference/contrib/settings.md:57]: Updated in https://github.com/wagtail/wagtail/pull/9390/commits/0a7f698a51a348f1790eb786c08f3f5a125d928c

File: wagtail/admin/templatetags/wagtailadmin_tags.py
Comment (Tijani-Dia) [reply wagtail/admin/templatetags/wagtailadmin_tags.py]: https://github.com/wagtail/wagtail/pull/9390/commits/73052d0d83ed3f12fc02dab1cc6caf2adc376e9d

File: wagtail/contrib/settings/templates/wagtailsettings/shared/header.html
Comment (Tijani-Dia) [reply wagtail/contrib/settings/templates/wagtailsettings/shared/header.html]: This is great, thanks! Updated in https://github.com/wagtail/wagtail/pull/9390/commits/d0be483510fa69a1569b497a425921a5751e219b

File: wagtail/contrib/settings/tests/translatable_generic/test_admin.py
Comment (zerolab) [reply wagtail/contrib/settings/tests/translatable_generic/test_admin.py:261]: AH, yep. Revisiting this with a rested head, it makes sense.

File: wagtail/contrib/settings/models.py
Comment (kaedroho) [reply wagtail/contrib/settings/models.py:272]: The reason for splitting them was because of the unique key on `site`. Introducing locales would mean we would need multiple settings per site, one for each locale.

Comment (zerolab) [reply wagtail/contrib/settings/models.py:272]: That would be the expected behaviour, though. No?
with `BaseTranslatableSiteSetting` you still end up with multiple settings per site:

- Site A, English
- Site A, French
- Site B, English
etc.

Comment (kaedroho) [reply wagtail/contrib/settings/models.py:272]: Yep, but if you added `TranslatableMixin` to a `BaseSiteSetting`, the unique key on the `site` field would prevent Site A from having two settings objects

Comment (zerolab) [reply wagtail/contrib/settings/models.py:272]: I get that. What I am suggesting here is that we add `unique_together = [("site", "locale"), ("translation_key", "locale")]` to `BaseSiteSetting` as it is done for `BaseTranslatableSiteSetting`, thus removing the need for `BaseTranslatableSiteSetting`

Comment (Tijani-Dia) [reply wagtail/contrib/settings/models.py:272]: Should I go ahead and unify the classes then?

Comment (zerolab) [reply wagtail/contrib/settings/models.py:272]: My preference would be to unify them. @kaedroho?

Comment (ababic) [reply wagtail/contrib/settings/models.py:272]: @zerolab With respect, this work was sponsored by a client, and it so happens that they have a combination of settings models... some of which require translation, and others that do not. For example, there are `DonationSettings` and `MarketingSignupSettings` that contain only configurational (non user-facing) things that do not (and should not) change on a per-language basis. There are others too, such as `SocialMediaLinks`, where the values ARE user-facing, but are simply a series of URLFields, where the same values apply for all locales. Another example would be GA settings.

In my view, Pages are a different kettle of fish entirely, so there's no need to emulate what happens there. Settings usage, in my experience, tends to be much similar to Snippets, where it's also perfectly valid to have both translatable and non-translatable models. Developers generally just use Settings over Snippets to get the 'singleton-like' interface and easy access in templates.

File: wagtail/contrib/settings/models.py
Comment (ababic) [inline wagtail/contrib/settings/models.py:120]: I do think there are too many classes here now. Can we not keep this as `BaseSiteSetting`, and subclass that to define `BaseTranslatableSiteSetting`? The less prohibitive `site` field on `BaseTranslatableSiteSetting`  should take precedence when extended to create a concrete model.

File: wagtail/contrib/settings/utils.py
Comment (ababic) [reply wagtail/contrib/settings/utils.py:56]: @Tijani-Dia Please see my comments on the backport MR in GltLab RE these utils. I understand the aim is to pull out some repetitive code to make things in the views a little more readable (and I admire that), but I think this is the 'wrong abstraction'. We don't actually ever want to get a Locale for a `model`, because that isn't meaningful or possible. And when we do want to know the Locale for a `request`, not only does that not really need to have anything to do with `model`, but the fact that we derive it from `request.GET['locale']` is something that is really quite specific to these views, and not really something that can be reused elsewhere.

File: wagtail/contrib/settings/models.py
Comment (ababic) [inline wagtail/contrib/settings/models.py:296]: This will not find the alternative version if target items have `translation_key` values that don't fit our 'fixed formula' - which I don't think we really have any grounds to impose... the migration path for existing settings models will likely be very similar to that of Snippet models, where `translation_key` values are first set in a data migration, and are just randomly generated at that point. We need to support that.

`TranslatableMixin` has an existing solution for getting the appropriate translation in this case: the `localized` property, and it doesn't require access to the request or a separately provided `Locale` object. We should find a way to utlise this here and try to keep the rest of the API simple..

File: wagtail/contrib/settings/models.py
Comment (zerolab) [reply wagtail/contrib/settings/models.py:272]: @ababic I am less concerned about something being sponsored by a client per se as we want to make it work for everyone.

Thank you for the additional example -- yep, that is a real and valid use case that I did not have at the back of my mind last week. This is the type of detail PR descriptions could really do with.

Happy for them to be separate on that basis

File: wagtail/contrib/settings/models.py
Comment (ababic) [reply wagtail/contrib/settings/models.py:296]: FYI, here's what worked in the backported version - a simple override to `get_site()`:

```
    @classmethod
    def for_site(cls, site, locale=None):
        """
        Overrides BaseSiteSetting.for_site() to get or create settings for the
        'provided' locale when editing OR the 'active' locale when retrieving
        an instance to display to users.
        """
        instance, created = cls.base_queryset().get_or_create(
            site=site,
            locale=locale or Locale.get_active(),
            defaults={"translation_key": get_default_translation_key(site)},
        )
        return instance
```

File: wagtail/contrib/settings/utils.py
Comment (zerolab) [reply wagtail/contrib/settings/utils.py:56]: Great shout @ababic. Always good to have a fresh pair of eyes

aside: I left the decision on this for the second review pass as it was a large enough PR to go through last time. Hope to come back this week with afresh, but it didn't happen.

File: wagtail/contrib/settings/models.py
Comment (ababic) [reply wagtail/contrib/settings/models.py:296]: On second thoughts. I don't think we really do want to automatically create versions in the active language when requesting from the front-end, so I have evolved the above to the following in a client project:

```
@classmethod
def for_site(cls, site, locale=None):
    ideal_translation_key = get_default_translation_key(site)

    if locale:
        # For requests from the admin, we do always want to
        # get or create an instance in the supplied locale
        instance, _ = cls.base_queryset().get_or_create(
            site=site,
            locale=locale,
            defaults={"translation_key": ideal_tranlsation_key},
        )

    # For requests from the front-end, we do not want to automatically
    # create objects in the selected language when they do not yet exist
    # In those cases: We want to return the default language version
    # instead (or create that if it doesn't exist)

    active_locale = Locale.get_active()
    default_locale = Locale.get_default()
    default_language_version = None

    for instance in cls.base_queryset().filter(
        site=site,
        locale__in=[
            active_locale,
            default_locale,
        ],
    ):
        if instance.locale_id == active_locale.id:
            # Hoorah! we have our preferred return value
            return instance
        else:
            default_language_version = instance

    # Return the default language version if it exists
    if default_language_version:
        return default_language_version

    # Or create it if it doesn't
    return cls.base_queryset().create(
        site=site,
        locale=default_locale,
        translation_key=ideal_translation_key,
    )

```

File: wagtail/contrib/settings/utils.py
Comment (Tijani-Dia) [reply wagtail/contrib/settings/utils.py:56]: @ababic I mainly copied the logic in the `LocaleMixin` here: https://github.com/wagtail/wagtail/blob/main/wagtail/admin/views/generic/mixins.py#L87, it checks for both the model and the `locale` parameter in the `GET` request.

Comment (ababic) [reply wagtail/contrib/settings/utils.py:56]: @Tijani-Dia I see. From a design perpective, this works as a view method because all of the data is accessible from the current view instance. However, it doesn't translate very well to a separate function (where everything must be provided as an argument). Do you see what I mean? It's actually adding unnecessary complication in a couple of cases.

File: wagtail/contrib/settings/models.py
Comment (ababic) [reply wagtail/contrib/settings/models.py:296]: On reflection, a lot of this complication comes from the fact the view is using this method, when it doesn't really have to. We can dedicate this method to finding an object for the front-end to use, and shift the rest into the Admin view.

**Comment by lb-** [comment]
Hey @Tijani-Dia what's the latest on this. Lots of comments and many seem to have been addressed. 

Does this need another review? Are you able to rebase / resolve conflicts and force push.

It would be great to see this in, thanks so much for processing a lot of the feedback so far.

**Comment by Tijani-Dia** [comment]
Hey @lb- , as Andy noted, we were doing this as part of client work that's now complete. I haven't got spare time to look into it since.

I'll need to address some of Andy's points before another review. I'll try to do it in the next weeks.

**Comment by lb-** [comment]
Really appreciate it - thank you Tidiane

**Comment by Tijani-Dia** [comment]
Hey @SteinRobert (cc @lb-), I haven't picked this back since @ababic's comments. At this point, I'd say it's very unlikely that I'll continue working on this - at least in the short term. If someone is willing to work on it, I'd recommend opening a new PR taking into accounts the comments made here.

**Comment by lb-** [comment]
This came up again in the Slack chat, I'm thinking it may be best to raise an enhancement request for this feature. From there we can see what other community members may be willing to work through this PR to get the feature over the line.

We can also close this PR, as indicated above, it's unlikely to be continued as is.

**Comment by lb-** [comment]
Closing this now, an issue has been raised with links to this and the other PRs.

https://github.com/wagtail/wagtail/issues/11508

Also, it's good to note that the `EditView` is already a class based view now https://github.com/wagtail/wagtail/pull/9317/files

**Pull request closed without merging**
